subprojects/publish/src/main/java/org/gradle/api/publish/internal/GradleModuleMetadataWriter.java

```diff
@@ -573,4 +584,83 @@ private ComponentData(ModuleVersionIdentifier coordinates, ImmutableAttributes a
             this.attributes = attributes;
         }
     }
+
+    @NotThreadSafe
+    public static class InvalidPublicationChecker {
+        private final String publicationName;
+        private final BiMap<String, VariantIdentity> variants = HashBiMap.create();
+        private List<String> errors;
+
+        public InvalidPublicationChecker(String publicationName) {
+            this.publicationName = publicationName;
+        }
+
+        public void registerVariant(String name, AttributeContainer attributes, Set<? extends Capability> capabilities) {
+            if (attributes.isEmpty()) {
+                failWith("Variant '" + name + "' must declare at least one attribute.");
+            }
+            if (variants.containsKey(name)) {
+                failWith("it's invalid to have 2 variants with the same name ('" + name + "')");
+            } else {
+                VariantIdentity identity = new VariantIdentity(attributes, capabilities);
+                if (variants.containsValue(identity)) {
+                    String found = variants.inverse().get(identity);
+                    failWith("variants '" + found + "' and '" + name + "' have the same attributes and capabilities. Please make sure either attributes or capabilities are different.");
```

Comment:
:o: "variants" -> "Variants" (start with upper case)

Reply:
Nitpick "Please make sure either attributes or capabilities are different" -> "Please make sure attributes and/or capabilities differ between variants."

Reply:
I'm not convinced the rephrase is clearer :)

Reply:
ok :)